File: feature_tests/dotnet/Tests/Option.cs
using System;

using Xunit;

namespace DiplomatFeatures.Tests;

public class OptionTests
{
    [Fact]
    public void NewWithValue()
    {
        OptionOpaque? o = OptionOpaque.New(1415);
        Assert.NotNull(o);
        if (o != null) // <-- to suppress warning
        {
            o.AssertInteger(1415);
        }
    }

    [Fact]
    public void NewNoneReturnsNone()
    {
        OptionOpaque? o = OptionOpaque.NewNone();
        Assert.Null(o);
    }

    [Fact]
    public void NewStruct()
    {
        OptionStruct o = OptionOpaque.NewStruct();

        Assert.Equal((ulong)904, o.C);
        // TODO: managed custom types in non-opaque struct fields

        // Using user-provided methods
        o.AssertIntegerForA(101);
        o.AssertCharForB(0x9910);
        o.AssertIntegerForD(926535);
    }

    [Fact]
    public void NewStructNones()
    {
        OptionStruct o = OptionOpaque.NewStructNones();

        Assert.Equal((ulong)908, o.C);
        // TODO: managed custom types in non-opaque struct fields

        // Using user-provided methods
        Assert.True(o.AIsNull());
        Assert.True(o.BIsNull());
        Assert.True(o.DIsNull());
    }
}

File: feature_tests/js2/test/option.mjs
import test from 'ava';

import { OptionOpaque } from "diplomat-wasm-js2-feature-tests";

test("Verify option methods", t => {
    const o = OptionOpaque.new(5);
    o.assert_integer(5);

    const on = OptionOpaque.new_none();
    t.assert(!on);

    const s = OptionOpaque.new_struct();

    s.a.assert_integer(101);
    s.b.assert_char('餐');
    t.is(s.c, 904);
    s.d.assert_integer(926535);

    const sn = OptionOpaque.new_struct_nones();
    t.assert(!sn.a);
    t.assert(!sn.b);
    t.is(sn.c, 908);
    t.assert(!sn.d);
});
